### PR TITLE
refactor(payment_methods): unify locker api function call

### PR DESCRIPTION
**Pull request by Chethan-rao**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR performs below changes - 
* Creates a single method using which locker api can be called and deserialised to the given type. 
* Currently, if locker returns an Err, the error response is not JWE+JWS decrypted. Instead its logged directly. This new method calls the api and irrespective of whether locker responds with an OK response or Err response, JWE+JWS decryption happens on the locker response. 
* Sends `x-tenant-id` in the headers to locker

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

To test, I failed the locker on purpose and received below error report in Hyperswitch which also includes JWE+JWS decrypted error message.

image

Sandbox testing - Basic sanity tests which also involves saving card in locker would suffice 

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payment_methods/cards.rs [hunk 1]
```diff
@@ -2204,6 +2196,59 @@ pub async fn call_to_locker_hs(
...
+    let locker = &state.conf.locker;
+    let jwekey = state.conf.jwekey.get_inner();
+    let response_type_name = type_name!(T);
+
+    let response = services::call_connector_api(state, request, flow_name)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .change_context(errors::VaultError::ApiError)?;
+
+    let is_locker_call_succeeded = response.is_ok();
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/cards.rs:2218]: Can you add comments to elaborate this behavior.

File: crates/router/src/core/payment_methods/cards.rs [hunk 2]
```diff
@@ -2204,6 +2196,62 @@ pub async fn call_to_locker_hs(
...
+    T: serde::de::DeserializeOwned,
+{
+    let locker = &state.conf.locker;
+    let jwekey = state.conf.jwekey.get_inner();
+    let response_type_name = type_name!(T);
+
+    let response = services::call_connector_api(state, request, flow_name)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .change_context(errors::VaultError::ApiError)?;
```
Comment (SanchithHegde) [inline crates/router/src/core/payment_methods/cards.rs:2216]: Why are we switching context to an internal server error and then immediately switching to `VaultError`?

File: crates/router/src/core/payment_methods/transformers.rs [hunk 3]
```diff
@@ -319,6 +320,7 @@ pub async fn mk_add_locker_request_hs(
     url.push_str("/cards/add");
     let mut request = services::Request::new(services::Method::Post, &url);
     request.add_header(headers::CONTENT_TYPE, "application/json".into());
+    request.add_header(headers::X_TENANT_ID, tenant_id.into());
```
Comment (SanchithHegde) [inline crates/router/src/core/payment_methods/transformers.rs:323]: If it's not asking for too much, can we have our request ID sent to the locker in a header and have the locker include it in its logs, so that it helps debugging?

Can be taken up in a separate PR.

File: crates/router/src/core/payment_methods/cards.rs (hunk 2 above)
Comment (Chethan-rao) [reply crates/router/src/core/payment_methods/cards.rs:2216]: We can directly raise VaultError, will remove ApiErrorResponse

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/core/payment_methods/transformers.rs [hunk 4]
```diff
@@ -319,6 +322,13 @@ pub async fn mk_add_locker_request_hs(
...
+    if let Some(req_id) = request_id {
+        request.add_header(
+            headers::X_REQUEST_ID,
+            req_id.as_hyphenated().to_string().as_str().into(),
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:329]: `as_str` is not required.

`crates/masking/src/maskable.rs:108:21`

File: crates/router/src/core/payment_methods/transformers.rs [hunk 5]
```diff
@@ -458,6 +471,14 @@ pub async fn mk_get_card_request_hs(
     url.push_str("/cards/retrieve");
     let mut request = services::Request::new(services::Method::Post, &url);
     request.add_header(headers::CONTENT_TYPE, "application/json".into());
+    request.add_header(headers::X_TENANT_ID, tenant_id.into());
+    if let Some(req_id) = request_id {
+        request.add_header(
+            headers::X_REQUEST_ID,
+            req_id.as_hyphenated().to_string().as_str().into(),
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:478]: Same here

File: crates/router/src/core/payment_methods/transformers.rs [hunk 6]
```diff
@@ -527,6 +550,14 @@ pub async fn mk_delete_card_request_hs(
     url.push_str("/cards/delete");
     let mut request = services::Request::new(services::Method::Post, &url);
     request.add_header(headers::CONTENT_TYPE, "application/json".into());
+    request.add_header(headers::X_TENANT_ID, tenant_id.into());
+    if let Some(req_id) = request_id {
+        request.add_header(
+            headers::X_REQUEST_ID,
+            req_id.as_hyphenated().to_string().as_str().into(),
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:557]: +1

File: crates/router/src/core/payment_methods/transformers.rs [hunk 7]
```diff
@@ -563,6 +596,14 @@ pub async fn mk_delete_card_request_hs_by_id(
...
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:603]: +1

File: crates/router/src/core/payment_methods/transformers.rs [hunk 8]
```diff
@@ -641,12 +682,22 @@ pub fn mk_crud_locker_request(
...
     let mut url = locker.basilisk_host.to_owned();
     url.push_str(path);
     let mut request = services::Request::new(services::Method::Post, &url);
     request.add_default_headers();
     request.add_header(headers::CONTENT_TYPE, "application/json".into());
+    request.add_header(headers::X_TENANT_ID, tenant_id.into());
+    if let Some(req_id) = request_id {
+        request.add_header(
+            headers::X_REQUEST_ID,
+            req_id.as_hyphenated().to_string().as_str().into(),
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:697]: +1

**Review by NishantJoshi00 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Review by ShankarSinghC — APPROVED** [review]
